scripts/update_function_docs/update_function_docs.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Usage: update_function_docs <RELEASE_BRANCH>
+//
+// e.g. update_function_docs origin/apply-setters/v0.2
+//
+// The command will checkout the release branch and update the function/example
+// docs with the latest patch version for the release. If the docs are updated
+// then a commit is created with the changes. The manual steps left to the user
+// are to push the commit to a branch and create a pull request.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"golang.org/x/mod/semver"
+	"gopkg.in/yaml.v2"
+)
+
+func exitWithErr(err error) {
+	fmt.Fprintf(os.Stderr, "%v\n", err)
+	os.Exit(1)
+}
+
+func runCmd(name string, arg ...string) (string, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.Command(name, arg...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	fmt.Printf("%s\n", cmd.String())
+	err := cmd.Run()
+	if err != nil {
+		return stdout.String(), fmt.Errorf("%s\n%s", stderr.String(), err)
+	}
+	return stdout.String(), err
+}
+
+func isCleanRepo() bool {
+	_, err := runCmd("git", "diff-index", "--quiet", "HEAD", "--")
```

Comment:
Why is it better than running `git status`?

Reply:
The primary reason is that this is less fragile than the `git status` alternative.

`git status` returns a zero exit code in both scenarios:
- No changes between HEAD and working tree (i.e. clean repo)
- Changes between HEAD and working tree (i.e. dirty repo)

The only way to differentiate between the two using the porcelain `git status` would be to parse stdout/stderr.

However, the plumbing `git diff-index` command offers an alternative approach - as it can be configured to return a nonzero exit code if there are changes between HEAD and the working tree.

See: https://git-scm.com/docs/git-diff-index#Documentation/git-diff-index.txt---quiet